pkg/sidecar/sidecar.go

```diff
@@ -135,45 +157,66 @@ func newWorkloadAPIClient(agentAddress string, timeout time.Duration) workload.X
 	return workload.NewX509Client(config)
 }
 
-//signalProcess sends the configured Renew signal to the process running the proxy
-//to reload itself so that the proxy uses the new SVID
-func (s *sidecar) signalProcess() (err error) {
-	if atomic.LoadInt32(&s.processRunning) == 0 {
-		cmd := exec.Command(s.config.Cmd, strings.Split(s.config.CmdArgs, " ")...)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		err = cmd.Start()
-		if err != nil {
-			return fmt.Errorf("error executing process: %v\n%v", s.config.Cmd, err)
-		}
-		s.process = cmd.Process
-		go s.checkProcessExit()
-	} else {
-		// Signal to reload certs
-		sig, err := getSignal(s.config.RenewSignal)
-		if err != nil {
-			return fmt.Errorf("error getting signal: %v\n%v", s.config.RenewSignal, err)
+// signalProcess sends the configured Renew signal to the process running the proxy
+// to reload itself so that the proxy uses the new SVID
+func (s *Sidecar) signalProcess() (err error) {
+	switch s.config.ExternalProcess {
+	case false:
+		if atomic.LoadInt32(&s.processRunning) == 0 {
+			cmd := exec.Command(s.config.Cmd, strings.Split(s.config.CmdArgs, " ")...) // #nosec
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			err = cmd.Start()
+			if err != nil {
+				return fmt.Errorf("error executing process: %v\n%v", s.config.Cmd, err)
+			}
+			s.process = cmd.Process
+			go s.checkProcessExit()
+		} else {
+			// Signal to reload certs
+			sig := unix.SignalNum(s.config.RenewSignal)
+			if sig == 0 {
+				return fmt.Errorf("error getting signal: %v", s.config.RenewSignal)
+			}
+
+			err = s.process.Signal(sig)
+			if err != nil {
+				return fmt.Errorf("error signaling process with signal: %v\n%v", sig, err)
+			}
 		}
 
-		err = s.process.Signal(sig)
-		if err != nil {
-			return fmt.Errorf("error signaling process with signal: %v\n%v", sig, err)
+	case true:
+		if atomic.LoadInt32(&s.processRunning) == 1 {
```

Comment:
It seems a little weird for an outside caller to have to educate the library about whether or not an externally managed process is running (via ProcessStarted) just so the library can know whether or not it should invoke the caller supplied callback.

I think things would be much simpler (and require less outside coordination with the library) if the invocation of ProcessReloadFunc didn't care about `s.processRunning` (and consequently `ProcessStarted` can be removed. The outside caller will already need to be tracking the lifetime of the external process and so can easily decide to ignore a ProcessReloadFunc callback if the process isn't running.

Reply:
done